runtime/browser/xwalk_content_base.cc

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2015 Intel Corporation. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "xwalk/runtime/browser/xwalk_content_base.h"
+
+#include <string>
+
+#include "base/prefs/pref_service.h"
+#include "components/autofill/content/browser/content_autofill_driver_factory.h"
+#include "components/autofill/core/browser/autofill_manager.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/render_view_host.h"
+#include "content/public/common/renderer_preferences.h"
+#include "xwalk/runtime/browser/xwalk_autofill_client.h"
+#include "xwalk/runtime/browser/xwalk_content_browser_client.h"
+#include "xwalk/runtime/browser/xwalk_browser_context.h"
+#include "xwalk/runtime/browser/xwalk_runner.h"
+
+#if defined(OS_ANDROID)
+#include "base/android/locale_utils.h"
+#include "xwalk/runtime/browser/android/xwalk_autofill_client_android.h"
+#include "xwalk/runtime/browser/android/xwalk_content.h"
+#endif
+
+#if !defined(OS_ANDROID)
+#include "xwalk/runtime/browser/xwalk_autofill_client_desktop.h"
+#endif
+
+namespace xwalk {
+
+const void* kXWalkContentUserDataKey = &kXWalkContentUserDataKey;
+
+class XWalkContentUserData : public base::SupportsUserData::Data {
+ public:
+  explicit XWalkContentUserData(XWalkContentBase* ptr) : content_(ptr) {}
+  static XWalkContentBase* GetContents(content::WebContents* web_contents) {
+    if (!web_contents)
+      return nullptr;
+    XWalkContentUserData* data = reinterpret_cast<XWalkContentUserData*>(
+      web_contents->GetUserData(kXWalkContentUserDataKey));
+    return data ? data->content_ : nullptr;
+  }
+
+ private:
+  XWalkContentBase* content_;
```

Comment:
class XWalkAutofillClientAndroid
    : public XWalkAutofillClient,
      public content::WebContentsUserData<XWalkAutofillClientAndroid> (and for desktop as well)

Then we set XWalkContentUserData as the user data on the web content which is used later to recover saved data : see various calls of  XWalkAutofillClient* autofill_manager_delegate =
      XWalkAutofillClientAndroid::FromWebContents(contents); -> which goes into WebContentsUserData